crates/hyvectl/src/out.rs

```diff
@@ -0,0 +1,177 @@
+use crate::color::Theme;
+use indicatif::ProgressBar;
+use serde::Serialize;
+use std::io::{Result, Write};
+
+#[derive(Clone, Debug, Serialize)]
+pub enum CommandOutput {
+    Message(String),
+    Result {
+        fields: Vec<(&'static str, String)>,
+        #[serde(skip_serializing)]
+        tty_template: String,
+        #[serde(skip_serializing)]
+        non_tty_template: String,
+    },
+    Progress(u64),
+    Spinner {
+        tick_strings: Vec<String>,
+        message: String,
+    },
+}
+
+impl CommandOutput {
+    pub fn message(message: &str) -> Self {
+        Self::Message(message.into())
+    }
+
+    pub fn result() -> Self {
+        Self::Result {
+            fields: vec![],
+            tty_template: String::default(),
+            non_tty_template: String::default(),
+        }
+    }
+
+    pub fn progress(progress: u64) -> Self {
+        Self::Progress(progress)
+    }
+
+    pub fn spinner(message: &str, tick_strings: &[&str]) -> Self {
+        Self::Spinner {
+            tick_strings: tick_strings.iter().map(|s| (*s).to_string()).collect(),
+            message: message.to_string(),
+        }
+    }
+
+    pub fn with_field(mut self, key: &'static str, value: String) -> Self {
+        if let Self::Result { fields, .. } = &mut self {
+            fields.push((key, value));
+        }
+        self
+    }
+
+    pub fn with_tty_template(mut self, template: &'static str) -> Self {
+        if let Self::Result { tty_template, .. } = &mut self {
+            *tty_template = template.into();
+        }
+        self
+    }
+
+    pub fn with_non_tty_template(mut self, template: &'static str) -> Self {
+        if let Self::Result {
+            non_tty_template, ..
+        } = &mut self
+        {
+            *non_tty_template = template.into();
+        }
+        self
+    }
+
+    fn to_json(&self) -> Result<String> {
+        use serde_json::{json, Map, Value};
+
+        match &self {
+            Self::Result { fields, .. } => {
+                let mut obj = Map::new();
+
+                for (key, field) in fields {
+                    let val: Value = json!(field);
+                    obj.insert((*key).to_string(), val);
+                }
+
+                let s = serde_json::to_string_pretty(&Value::Object(obj))?;
```

Comment:
I would also provide different outputs here depending on whether we're in a TTY or not. E.g. non-tty output should be `serve_json::to_string` (so without any formatting) and TTY output could be `colored_json::write_colored_json` (needs the `colored_json` dependency but looks very nice in return).